Use the same criteria to flip chiral atoms in SMILES parsing/writing (Fixes #1028)

ricrogz opened the pull request:
Explanation of the problem is in the comments of Issue #1028 . But it summarizes into that SMILES parsing and writing were using different criteria to decide if the chiral tag of an atom had to be flipped or not.

This PR moves the code to decide whether to flip a chiral atom into Canon.cpp, and uses the same function to check chiral atoms at writing and parsing of SMILES strings.

I have not checked the change in the python test, but both strings seem to be "stable" and not flip on roundtripping the SMILES -- if the fix of the reference is wrong, then maybe the algorithm to decide the flipping is not quite correct.

greglandrum commented:
Thanks for digging into this and fixing a bug I've been avoiding for far too long. :-)

Just for completeness, please add a test case for the example from #1028 as well: `O[C@H]1CC2CCC(C1)[N@@]2C`

greglandrum commented on Code/GraphMol/Wrap/rough_test.py:
```diff
@@ -3771,7 +3771,7 @@ def testSequenceBasics(self):
     seq = 'CYIQNCPLG'
     fasta = '>\nCYIQNCPLG\n'
-    smi = 'CC[C@H](C)[C@H](NC(=O)[C@H](Cc1ccc(O)cc1)NC(=O)[C@@H](N)CS)C(=O)N[C@@H](CCC(N)=O)C(=O)N[C@@H](CC(N)=O)C(=O)N[C@@H](CS)C(=O)N1CCC[C@H]1C(=O)N[C@@H](CC(C)C)C(=O)NCC(=O)O'
+    smi = 'CC[C@H](C)[C@H](NC(=O)[C@H](Cc1ccc(O)cc1)NC(=O)[C@@H](N)CS)C(=O)N[C@@H](CCC(N)=O)C(=O)N[C@@H](CC(N)=O)C(=O)N[C@@H](CS)C(=O)N1CCC[C@@H]1C(=O)N[C@@H](CC(C)C)C(=O)NCC(=O)O'
```
I'm pretty sure that the original SMILES here was correct and that you've inverted the stereochemistry of the proline.

ricrogz replied:
You are right, it was ok. Apparently, at the time of deciding if the chiral center should be flipped, it did not have an explicit H, but an implicit one, and ended up being flipped.

I have added the implicit valence to the flipping decision, which seems to work fine, and solves this problem.

ricrogz commented:
> Thanks for digging into this and fixing a bug I've been avoiding for far too long. :-)
> 
> Just for completeness, please add a test case for the example from #1028 as well: `O[C@H]1CC2CCC(C1)[N@@]2C`

No problem, Greg. It was quite entertaining to look into this :)

I added the molecule from the issue, and fixed the issue with the sequence by also considering the implicit value when making the decision to flip a chiral center.

greglandrum commented on Code/GraphMol/Wrap/rough_test.py:
```diff
@@ -8,19 +8,25 @@
 # Boost functions are NOT found by doctest, this "fixes" them
 #  by adding the doctests to a fake module
 import importlib.util
+import os
```
I understand why the automatic code formatter did this, but it really made the code less readable.
Please undo this change (so that the `import importlib.util` line and the comment before it are directly above line 30

ricrogz replied:
Oh! I'm sorry, I didn't even notice it did that. I just reverted it.
